### DISCUSSION TITLE
How to run my own crawler to use the Docsearch plugin?

### DISCUSSION THREAD

**Discussion by Yue-plus** [vuepress/core#92] [Q&A] [answered]
How to run my own crawler to use the Docsearch plugin?
如何运行自己的爬虫来使用 Docsearch 插件？

------

# Symptom description / 症状描述

I added the `@vuepress/plugin-docsearch` plugin to my VuePress2 project and ran my own crawler by referring to the "[Running Your Own Crawler](https://docsearch.algolia.com/docs/run-your-own/)" document.
我在自己的 VuePress2 项目中添加了 `@vuepress/plugin-docsearch` 插件，并参照”[运行你自己的爬虫](https://docsearch.algolia.com/docs/run-your-own/)“ 文档来运行了自己的爬虫。

Now there are records in Indices:
现在 Indices 中已经有 records 了：

![image](https://user-images.githubusercontent.com/33195150/113150383-99d31880-9266-11eb-8669-1836c11e147f.png)

But no results are found in VuePress2:
但在 VuePress2 中搜索不到任何结果：

![image](https://user-images.githubusercontent.com/33195150/113151404-9e4c0100-9267-11eb-941d-50cc79b6301a.png)

Project link: <https://note.yueplus.ink/>
项目地址：<https://note.yueplus.ink/>

-------

# Operation process / 操作过程

## VuePress configuration / VuePress 配置

Add `@vuepress/plugin-docsearch` plugin:
添加  `@vuepress/plugin-docsearch` 插件：

```bash
yarn add @vuepress/plugin-docsearch@next
```

`package.json`:

```json
{
  "devDependencies": {
    "vuepress": "^2.0.0-beta.6"
  },
  "dependencies": {
    "@vuepress/plugin-docsearch": "^2.0.0-beta.6",
  }
}
```

`.vuepress/config.ts`:

```ts
  plugins: [
    ['@vuepress/docsearch', {
      appId: '0YWM7BGDQI',                     // Application ID
      apiKey: '9dc6fe12516690be51dea370e78001f6',   // Search-Only API Key
      indexName: 'note.yueplus.ink',
      locales: {
        '/': {
          placeholder: '在此搜索文档'
        }
      }
    }]
  ]
```

## Run the crawl from the Docker image / 从 Docker 镜像运行爬虫

```
docker run --env-file=.env -e "CONFIG=$(cat /c/Users/Yue_plus/code/NetNote/docsearch.config.json | jq -r tostring)" algolia/docsearch-scraper
```

![image](https://user-images.githubusercontent.com/33195150/113154741-e587c100-926a-11eb-8cb5-4e3f7eac2c12.png)

`docsearch.config.json` :

```json
{
    "index_name": "note.yueplus.ink",
    "start_urls": ["https://note.yueplus.ink/"],
    "selectors": {
      "lvl0": ".page header h1",
      "lvl1": ".page article h1",
      "lvl2": ".page section h3",
      "lvl3": ".page section h4",
      "lvl4": ".page section h5",
      "lvl5": ".page section h6",
      "text": ".page header p,.page section p,.page section ol,.page ul li"
    }
  }
```

Maybe my `docsearch.config.json` file is incorrectly written?
可能是我的 `docsearch.config.json` 文件写的不正确？

**Comment by Yue-plus** [comment, marked as answer]
The data can be searched using the playground tool in [docsearch-scraper](https://github.com/algolia/docsearch-scraper), but the search results are not displayed normally:
使用 [docsearch-scraper](https://github.com/algolia/docsearch-scraper) 中的测试工具是可以搜索到数据的，但是却没有正常显示搜索结果：

![image](https://user-images.githubusercontent.com/33195150/113155739-eec55d80-926b-11eb-9e27-63e1e1a7f8a0.png)

**Reply by meteorlxy** [reply]
I think it's better to report or ask in docsearch project 🤔

**Reply by meteorlxy** [reply]
Seems like your config is wrong. Do you have `.page header h1` element in your site?

**Reply by meteorlxy** [reply]
In you screenshot, the content of `lvl0` to `lvl5` are `null`, which could be the reason

**Reply by Yue-plus** [reply]
> Seems like your config is wrong. Do you have `.page header h1` element in your site?

I'm using the default theme of VuePress2, can you provide a docsearch.config.json that can be used for the default theme in the [documentation](https://vuepress2.netlify.app/reference/plugin/docsearch.html#get-search-index)?
我正在使用 VuePress2 的默认主题，可否在 [文档](https://vuepress2.netlify.app/zh/reference/plugin/docsearch.html#%E8%8E%B7%E5%8F%96%E6%90%9C%E7%B4%A2%E7%B4%A2%E5%BC%95) 中提供一个可用于默认主题的 `docsearch.config.json`？

**Reply by Yue-plus** [reply]
When [Create a new configuration](https://docsearch.algolia.com/docs/run-your-own/#create-a-new-configuration), I encountered a problem:
参考 [创建新配置文件](https://docsearch.algolia.com/docs/run-your-own/#create-a-new-configuration) 操作时，遇到问题：

[An error was encountered while executing `./docsearch bootstrap`. #547](https://github.com/algolia/docsearch-scraper/issues/547)

**Reply by Yue-plus** [reply]
After changing the 'docsearch.config.json' file to the following:
将 `docsearch.config.json` 文件更改为如下内容后：

```json
{
  "index_name": "note.yueplus.ink",
  "start_urls": [
    "https://note.yueplus.ink/"
  ],
  "stop_urls": [],
  "selectors": {
    "lvl0": "main h1",
    "lvl1": "main h2",
    "lvl2": "main h3",
    "lvl3": "main h4",
    "lvl4": "main h5",
    "lvl5": "main h6",
    "text": "main p, main li"
  }
}
```

I cleared the Indices once, re-run the crawler, and crawled more records:
清空了一次 Indices，重新运行爬虫，爬取到了更多 records：

![image](https://user-images.githubusercontent.com/33195150/113289849-5a1e3680-9323-11eb-99fc-a05629010429.png)

The search results can now be displayed normally in the playground tool in [docsearch-scraper](https://github.com/algolia/docsearch-scraper) : 
现在能在 [docsearch-scraper](https://github.com/algolia/docsearch-scraper) 中的测试工具中正常展示搜索结果了：

![image](https://user-images.githubusercontent.com/33195150/113290033-a23d5900-9323-11eb-9dbc-962b772fb87c.png)

But now I still can’t search in VuePress2:
但现在任然无法在 VuePress2 中搜索：

![image](https://user-images.githubusercontent.com/33195150/113290482-49ba8b80-9324-11eb-8c18-f458f87535f7.png)

**Reply by meteorlxy** [reply]
You can refer to these configs:

- vuepress officail docs (both v1 and v2): https://github.com/algolia/docsearch-configs/blob/a797a5c2b1531174c403eb2ea46a8adefca92e6d/configs/vuepress.json
- vuepress-theme-reco: https://github.com/algolia/docsearch-configs/blob/a797a5c2b1531174c403eb2ea46a8adefca92e6d/configs/vuepress-theme-reco.json
- vuepress-theme-hope: https://github.com/algolia/docsearch-configs/blob/a797a5c2b1531174c403eb2ea46a8adefca92e6d/configs/vuepress-theme-hope.json

Some important parts:

```json
      "lang": {
        "selector": "/html/@lang",
        "type": "xpath",
        "global": true
      }
```

```json
  "custom_settings": {
    "attributesForFaceting": [
      "lang"
    ]
  },
```

I'll refine the docs later

**Reply by Yue-plus** [reply]
`docsearch.config.json`:

```json
{
  "index_name": "note.yueplus.ink",
  "start_urls": [
    "https://note.yueplus.ink/"
  ],
  "stop_urls": [],
  "selectors": {
    "lvl0": "main h1",
    "lvl1": "main h2",
    "lvl2": "main h3",
    "lvl3": "main h4",
    "lvl4": "main h5",
    "lvl5": "main h6",
    "text": "main p, main li",
    "lang": {
      "selector": "/html/@lang",
      "type": "xpath",
      "global": true
    }
  },
  "custom_settings": {
    "attributesForFaceting": ["lang", "tags"]
  }
}
```

![image](https://user-images.githubusercontent.com/33195150/113475019-d476c400-94a5-11eb-93bb-782f822fa7e7.png)

Oh Yeah!!!
好耶！！！

![image](https://user-images.githubusercontent.com/33195150/113475106-37685b00-94a6-11eb-9854-430d3aaf8c46.png)
